### PR TITLE
Lab 06

### PR DESCRIPTION
**Pull request by goremukin**

### CODE REVIEW

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+﻿using GodelMastery.Library.DataAccessLayer.Models;
+using GodelMastery.Library.DataAccessLayer.Repositories;
+using System;
+
+namespace GodelMastery.Library.BusinessLayer
+{
+    public class AuthorsService
+    {
+        private readonly IRepository<Author> authorsRepository;
+        public Author[] Authors => authorsRepository.ReadAll();
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: Array of authors - poor idea

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+﻿using GodelMastery.Library.DataAccessLayer.Models;
+using GodelMastery.Library.DataAccessLayer.Repositories;
+using System;
+
+namespace GodelMastery.Library.BusinessLayer
+{
+    public class AuthorsService
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: IAuthorsService

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public Author Find(int id)
+        {
+            return authorsRepository.Read(id);
+        }
+
+        public Author[] Find(Func<Author, bool> predicate)
+        {
+            return authorsRepository.Read(predicate);
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: `Where` the most suitable

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            authorsRepository.Create(author);
+            authorsRepository.Save();
+        }
+
+        public Author Find(int id)
+        {
+            return authorsRepository.Read(id);
+        }
+
+        public Author[] Find(Func<Author, bool> predicate)
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: Comment up

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        public void Remove(int id)
+        {
+            authorsRepository.Delete(id);
+            authorsRepository.Save();
+        }
+
+        public void Update(Author author)
+        {
+            authorsRepository.Update(author);
+            authorsRepository.Save();
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: 3 calls of Save method

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+namespace GodelMastery.Library.BusinessLayer
+{
+    public class BooksService
+    {
+        private const string CoversFolder = "~\\images\\covers";
+        private const string NoCoverFileName = "nocover.jpg";
+
+        private readonly IRepository<Book> booksRepository;
+        private readonly IRepository<Cover> coversRepository;
+        public Book[] Books => booksRepository.ReadAll();
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: the same as AuthorsService

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+using GodelMastery.Library.DataAccessLayer.Models;
+using GodelMastery.Library.DataAccessLayer.Repositories;
+using System.IO;
+using GodelMastery.Library.BusinessLayer.Helpers;
+using System.Drawing;
+using System.Drawing.Imaging;
+
+namespace GodelMastery.Library.BusinessLayer
+{
+    public class BooksService
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: the same as AuthorsService

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        
+        public BooksService(IRepository<Book> booksRepository, IRepository<Cover> coversRepository)
+        {
+            this.booksRepository = booksRepository;
+            this.coversRepository = coversRepository;
+        }
+
+        public static void DeleteCoverFile(Book book, string appRootPath)
+        {
+            if (book.Cover != null)
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: must be throw exc

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        public Book Find(int id)
+        {
+            return booksRepository.Read(id);
+        }
+
+        public Book[] Find(Func<Book, bool> predicate)
+        {
+            return booksRepository.Read(predicate);
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: the same as AuthorsService

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        public void DeleteCover(Cover cover)
+        {
+            coversRepository.Delete(cover.Id);
+            coversRepository.Save();
+        }
+
+        public void Update(Book book)
+        {
+            booksRepository.Update(book);
+            booksRepository.Save();
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: the same as AuthorsService (3 calls of Save)

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        {
+            ImageHelper helper = new ImageHelper(imageStream);
+            ImageFormat format;
+            try
+            {
+                format = helper.GetFormat();
+            }
+            catch
+            {
+                throw new ArgumentException("Incorrect image format.");
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: Invalid type of exc

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+        public void Update(Book book)
+        {
+            booksRepository.Update(book);
+            booksRepository.Save();
+        }
+
+        public Cover UploadCover(Stream imageStream, Book book, string appRootPath)
+        {
+            ImageHelper helper = new ImageHelper(imageStream);
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: you don't check the method parameters **anywhere**

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            }
+
+            if (ImageFormat.Jpeg.Equals(format))
+            {
+                extension = ".jpg";
+            }
+
+            if (string.IsNullOrEmpty(extension))
+            {
+                throw new ArgumentException("Incorrect image format.");
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: Invalid type of exc

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs [hunk 14]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using GodelMastery.Library.DataAccessLayer.Models;
+using GodelMastery.Library.DataAccessLayer.Repositories;
+
+namespace GodelMastery.Library.BusinessLayer
+{
+    public class GenresService
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs]: the same as BL services

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        private readonly IRepository<Genre> genresRepository;
+        public Genre[] Genres => genresRepository.ReadAll();
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs]: the same as BL services

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+        public Genre[] Find(Func<Genre, bool> predicate)
+        {
+            return genresRepository.Read(predicate);
+        }
+
+        public void Update(Genre genre)
+        {
+            genresRepository.Update(genre);
+            genresRepository.Save();
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs]: the same as BL services

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        public Genre Find(int id)
+        {
+            return genresRepository.Read(id);
+        }
+
+        public Genre[] Find(Func<Genre, bool> predicate)
+        {
+            return genresRepository.Read(predicate);
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs]: the same as BL services

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/Helpers/ImageHelper.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            Image image = null;
+            using (MemoryStream stream = new MemoryStream())
+            {
+                imageStream.CopyTo(stream);
+                image = Image.FromStream(stream);
+            }
+
+            var result = image.RawFormat;
+            image.Dispose();
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/Helpers/ImageHelper.cs]: `using`?

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Initializers/ApplicationDbInitializer.cs [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+            InitializeUsersAndBooks(context);
+        }
+
+        private void InitializeGenres(ApplicationDbContext context)
+        {
+            IList<Genre> genres = new List<Genre>();
+            genres.Add(new Genre { Name = "Fantasy" });
+            context.Genres.AddRange(genres);
+
+            context.Covers.Add(new Cover { Format = "jpg" });
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Initializers/ApplicationDbInitializer.cs]: cycle?

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+            }
+        }
+
+        public void Dispose()
+        {
+            if (disposed) return;
+            disposed = true;
+
+            Dispose(true);
+            GC.SuppressFinalize(this);
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs]: ` if (disposed) return;
disposed = true;`

wrong impl IDisposable

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            Dispose(true);
+        }
+
+        public TEntity Read(int id)
+        {
+            return dbSet.Find(id);
+        }
+
+        public TEntity[] Read(Func<TEntity, bool> predicate)
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs]: Wrong return type

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs (hunk 1 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs (hunk 6 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs (hunk 15 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs (hunk 21 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs (hunk 3 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+            }
+            else
+            {
+                ViewBag.Msg = "Successfully added.";
+            }
+
+            return Edit();
+        }
+
+        private bool FileSizeOverTheLimit(HttpPostedFileBase postedFile)
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs]: move to BL

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+using System.Web;
+using GodelMastery.Library.DataAccessLayer.Models;
+using Microsoft.AspNet.Identity;
+using System.Linq;
+
+namespace Lab06.MVC.Controllers
+{
+    public class BooksController : Controller
+    {
+        private const int FileSizeLimit = 5 * 1024 * 1024;
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs]: move to BL too

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+        {
+            var user = UserManager.FindById(User.Identity.GetUserId());
+            if (user != null)
+            {
+                return user.PasswordHash != null;
+            }
+            return false;
+        }
+
+        public enum ManageMessageId
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs]: move to BL

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+
+        private void AddErrors(IdentityResult result)
+        {
+            foreach (var error in result.Errors)
+            {
+                ModelState.AddModelError(string.Empty, error);
+            }
+        }
+
+        private bool HasPassword()
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs]: move to BL

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+        private async Task SignInAsync()
+        {
+            var user = await UserManager.FindByIdAsync(User.Identity.GetUserId());
+            if (user != null)
+            {
+                await SignInManager.SignInAsync(user, isPersistent: false, rememberBrowser: false);
+            }
+        }
+
+#region Helpers
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs]: move to services

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Models/IdentityModels.cs [hunk 27]
```diff
@@ -0,0 +1,18 @@
+﻿using GodelMastery.Library.DataAccessLayer.Models;
+using Microsoft.AspNet.Identity.EntityFramework;
+
+namespace Lab06.MVC.Models
+{
+    public class ApplicationDbContext : IdentityDbContext<ApplicationUser>
```
Comment (EdwOK) [inline Lab06.MVC/Lab06.MVC/Lab06.MVC/Models/IdentityModels.cs]: ApplicationDbContext already exists

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs (hunk 17 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs (hunk 2 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: All sevices in Lab6 project have been replaced by ISerive<>

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs (hunk 4 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/AuthorsService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs (hunk 7 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs (hunk 9 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs (hunk 11 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: replaced by BadImageFormatException

Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: replaced by BadImageFormatException

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs (hunk 8 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs (hunk 12 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: Added checking for null in constructors and in some methods

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs (hunk 13 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/BooksService.cs]: replaced by BadImageFormatException

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs (hunk 14 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/GenresService.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/Helpers/ImageHelper.cs (hunk 18 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.BusinessLayer/Helpers/ImageHelper.cs]: Sure, fixed

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+                    return RedirectToAction("Edit");
+                }
+            }
+
+            return Update(id);
+        }
+
+        [HttpPost]
+        [Authorize(Roles = "Administrator")]
+        public ActionResult Delete(int? id)
```
[the remark this answers is not in the document]
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs (hunk 20 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Repositories/Repository.cs]: fixed

File: Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Initializers/ApplicationDbInitializer.cs (hunk 19 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/GodelMastery.Library.DataAccessLayer/Initializers/ApplicationDbInitializer.cs]: fixed

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs (hunk 22 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs]: fixed

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs (hunk 23 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/BooksController.cs]: fixed

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs (hunk 26 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/Lab06.MVC/Controllers/ManageController.cs]: ManageMessageId enum and GetMessageString method moved to BL

File: Lab06.MVC/Lab06.MVC/Lab06.MVC/Models/IdentityModels.cs (hunk 27 above)
Comment (goremukin) [reply Lab06.MVC/Lab06.MVC/Lab06.MVC/Models/IdentityModels.cs]: fixed

**Pull request merged**